WXR_Parser_XML_Processor: An all-PHP WXR parser

adamziel opened the pull request:
## Description

Brings over the XMLProcessor class from [WordPress/php-toolkit](https://github.com/wordpress/php-toolkit) as a fallback XML parser. The end goal is removing the `WXR_Parser_Regex` class.

**This PR does not expose the new parser to the plugin consumers yet.** It only adds the new parser class and plugs it into the CI test pipeline.

## Rationale

`WXR_Parser_Regex` enables importing WXR files on hosts where libxml is not available. However, `WXR_Parser_Regex` cannot reliably parse XML documents. It does not conform to [the XML 1.0 spec](https://www.w3.org/TR/xml/) and fails multiple unit tests. `WXR_Parser_Regex` cannot be improved in small steps. Any iterative improvement strategy will ultimately lead to a full spec-compliant XML parser. That's what `XMLProcessor` is.

This PR proposes a new `WXR_Parser_XML_Processor` class that uses `XMLProcessor` as a backend. `XMLProcessor`:

* Understands XML nuances the regex-based parser can't process, e.g. namespaces, nesting, whitespaces, unicode.
* Is tested against the W3 XML test suite and is capable of stream-processing large XML files without running out of memory.
* Follows the ideas and conventions established in [WP_HTML_Processor](https://developer.wordpress.org/reference/classes/wp_html_processor/).

### Implementation details

This PR adds a copy of `WXREntityReader`, `XMLProcessor`, and their dependencies into this repository. This way:

* Cloning this repository is enough to use the `wordpress-importer` plugin, no `composer install` needed.
* `wordpress-importer` plugin remains small as it doesn't require pulling the entire `php-toolkit`.

In the future, it might be useful to switch to composer and find a way of pruning the installed dependencies.

### Follow-up work

* Measure the performance impact
* Bump PHP version checked by WPCS to 7.2 (same as latest WordPress core) to allow more static type annotations.
* Remove WXR_Parser_Regex.

cc @zaerl @dmsnell @sirreal @janjakes

dmsnell commented on src/parsers/xml-processor/class-xmlprocessor.php:
```diff
@@ -0,0 +1,4499 @@
+ *        * <!ATTLIST, see https://www.w3.org/TR/xml/#attdecls
+ *        * <!ENTITY, see https://www.w3.org/TR/xml/#sec-entity-decl
+ *        * <!NOTATION, see https://www.w3.org/TR/xml/#sec-entity-decl
+ *        * Conditional sections, see https://www.w3.org/TR/xml/#sec-condition-sect
```
skip over or bail? I guess if we ignore we can bail on an entity whose definition we don’t have.

adamziel replied:
We bail out on them. To ignore them, we'd need to be able to parse them. I'll update the comment to specify this more precisely.

adamziel replied:
I've updated the comment

dmsnell commented on src/parsers/xml-processor/class-xmlprocessor.php:
```diff
@@ -0,0 +1,4499 @@
+ * #### Special elements
+ *
+ * All XML elements are handled in the same way, except when you mark
+ * them as PCData elements. These are special because their contents
```
I think I’ve been wrong about this before on two accounts:

 - `PCDATA` is the normative text type we’re used to working with in HTML. tags inside PCDATA content are indeed tags, not plaintext.
 - `CDATA` is the verbatim text type, but an element cannot specify its content model as CDATA. an attribute’s content may be CDATA, but not an element’s.

this clears up the long confusion I had around differentiating content with a text span that looks like the closing tag of the wrapping CDATA element — it can’t happen. the only way to embed CDATA in an element is inside a CDATA section with `<![CDATA[…]]>`.

now I think we have some other concerns here with things like un-closed `post-content` elements. I’m directly inspired by the HTML parser’s stack machine for making decisions like “if there is an open `post-content` on the stack of open elements, generate implied end tags; otherwise ignore”

we might want or need this in the design, or announce it in a dry-run

adamziel replied:
Aha! This is really useful, thank you! I'll remove this entire semantics, then.

> now I think we have some other concerns here with things like un-closed post-content elements. 

No WXR importer can process that today, right? I'd love to eventually introduce an error token to put the API consumer in charge of what to do, though.

dmsnell commented on src/parsers/xml-processor/utf8.php:
```diff
@@ -0,0 +1,363 @@
+	$byte4 = chr( $code_point & 0x3F | 0x80 );
+
+	return "{$byte1}{$byte2}{$byte3}{$byte4}";
+}
```
resurrecting in [Core-63863](https://core.trac.wordpress.org/ticket/63863)

adamziel commented on src/parsers/xml-processor/class-wxrentityreader.php:
```diff
@@ -0,0 +1,984 @@
+ * * Speed – it should be as fast as possible
+ * * No PHP extensions required – it can run on any PHP installation
+ * * Reliability – no random crashes when encountering malformed XML or UTF-8 sequences
+ * * Low, predictable memory footprint to support 1000GB+ WXR files
```
I was shocked but they actually exist!

adamziel commented on src/parsers/xml-processor/class-xmlprocessor.php:
```diff
@@ -0,0 +1,4499 @@
+ * The final goal is to support both 1.0 and 1.1 depending on the
+ * initial processing instruction (<?xml version="1.0" ?>). We're
+ * starting with 1.0, however, because most that's what most WXR
+ * files declare.
```
Aha – really good idea! I'm applying these changes in the wordpress-components repo and will ping you there when I have a PR

adamziel replied:
Addressed in https://github.com/WordPress/php-toolkit/pull/150 and backported here

adamziel commented on src/parsers/xml-processor/class-wxrentityreader.php:
```diff
@@ -0,0 +1,984 @@
+ *
+ * `WP_WXR_Entity_Reader` would accumulate post data until the `wp:post_meta` tag. Then it would emit a
+ * `post` entity and accumulate the meta information until the `</wp:postmeta>` closer. Then it
+ * would advance to `<wp:post_id>` and **ignore it**.
```
We have `$xml_processor->bail()` that's inspired by the HTML API:

```php
			$this->bail(
				'Invalid closing tag encountered.',
				self::ERROR_SYNTAX
			);
```

It stores some contextual information that can be used later on to create a nice error message:

```php
		$this->exception  = new XMLUnsupportedException(
			$message,
			$this->get_token_type(),
			$starts_at,
			$token,
			$this->get_breadcrumbs()
		);
```

Now, your comment is more about higher-order concepts in the WXR file, like unrecognized entities. We don't have a standard way of announcing these at the moment. In the context of the `wordpress-importer` plugin, we'd need something that works with all the XML parsers shipped in this repo – `WXREntityReader` is specific to `XMLProcessor`.

One way to get there would be to update every WXR parser in here to provide meaningful error details in every error scenario. I'm nervous about this design as it requires consistent and correct reporting of the same errors across multiple parser implementations. I'm worried they'd diverge over time.

A safer way would be to rewire all the data flows in this repo from:

```
WXR_Parser_SimpleXML -> WP_Import
WXR_ParserXML -> WP_Import
XMLProcessor -> WXREntityReader -> WXR_Parser_XML_Processor -> WP_Import
```

To:

```
WXR_Parser_SimpleXML -> WXREntityReader -> WP_Import
WXR_ParserXML -> WXREntityReader -> WP_Import
XMLProcessor -> WXREntityReader -> WXR_Parser_XML_Processor -> WP_Import
```

And detect the errors at the WXREntityReader level. The XML parsing classes would be reduced to a consistent interface for walking through an XML document.

adamziel commented:
@dmsnell thank you for your great review! I've addressed most of your points in https://github.com/WordPress/php-toolkit/pull/149 and the rest of them inline.

adamziel commented on src/parsers/xml-processor/class-wxrentityreader.php:
```diff
@@ -0,0 +1,984 @@
+					'slug'        => $this->last_opener_attributes['nicename'],
+					'description' => $this->text_buffer,
+				);
+				$this->text_buffer            = '';
```
Thank you for bringing this up @dmsnell! I think we need to leave this up to the XMLProcesor consumer. Here's my reasoning:

ICU with it's Normalizer class is not always available and I'd hate to output different bytes for the same document in different runtimes.

[XML 1.0 spec](https://www.w3.org/TR/xml/) doesn't seem to be making any normalization recommendations. [XML 1.1](https://www.w3.org/TR/xml-c14n/), however, discusses normalization in section 4.2 No Character Model Normalization:

> The Unicode standard [[Unicode]](https://www.w3.org/TR/xml-c14n/#Unicode) allows multiple different representations of certain "precomposed characters" (a simple example is "ç"). Thus two XML documents with content that is equivalent for the purposes of most applications may contain differing character sequences. The W3C is preparing a normalized representation [[CharModel]](https://www.w3.org/TR/xml-c14n/#CharModel). The [C14N-20000119](https://www.w3.org/TR/xml-c14n/#C14N-20000119) Canonical XML draft used this normalized form. However, many **XML 1.0 processors do not perform this normalization**. Furthermore, applications that must solve this problem typically enforce character model normalization at all times starting when character content is created in order to avoid processing failures that could otherwise result (e.g. see example from [Cowan](https://www.w3.org/TR/xml-c14n/#CowanExample)). Therefore, character model normalization has been moved out of scope for XML canonicalization. However, the XML processor used to prepare the XPath data model input is required (by the [Data Model](https://www.w3.org/TR/xml-c14n/#DataModel)) to use Normalization Form C [[NFC](https://www.w3.org/TR/xml-c14n/#ref-NFC), [NFC-Corrigendum](https://www.w3.org/TR/xml-c14n/#NFC-Corrigendum)] when converting an XML document to the UCS character domain from any encoding that is not UCS-based (currently, UCS-based encodings include UTF-8, UTF-16, UTF-16BE, and UTF-16LE, UCS-2, and UCS-4).

The entity reader could be making note of unique normalized category names whenever ICU is available and making a recommendation to merge them.

adamziel commented on src/parsers/xml-processor/class-basebytereadstream.php:
```diff
@@ -0,0 +1,186 @@
+		}
+		$bytes                           = substr( $this->buffer, $this->offset_in_current_buffer, $n );
+		$this->offset_in_current_buffer += $n;
+		if ( $this->offset_in_current_buffer > $this->buffer_size ) {
```
Ditto the other comment about `$buffer_size`. It's now called `$max_lookbehind_bytes` and it only means the number of bytes we keep around to `seek()` back when needed – 2048 is typically enough and it can be adjusted for the scenario when more is needed.

adamziel commented on src/parsers/class-wxr-parser-xml-processor.php:
```diff
@@ -0,0 +1,372 @@
+		//
+		// Unfortunately, different WXR files defined the `wp` namespace in a different way.
+		// Folks use a mixture of HTTP vs HTTPS protocols and version numbers. We must
+		// account for all possible options to parse these documents correctly.
```
> I’m guessing there are subtle claimed differences between the given version numbers…

I'm guessing so too, they can probably be reverse-engineered by studying previous plugin releases.

adamziel commented:
@zaerl I'm really interested in your take on this PR

zaerl commented on src/parsers/xml-processor/utf8.php:
```diff
@@ -0,0 +1,363 @@
+	}
+
+	if ( $code_point <= 0x7FF ) {
+		$byte1 = chr( ( $code_point >> 6 ) | 0xC0 );
```
This is similar to [what I do](https://github.com/zaerl/mojibake/blob/main/src/encoding.c#L196), and others as well. What we have here is working because we are checking U+0080–U+07FF.

The mask I use is still considered a good practice: it documents that you want to “keep 5 bits” and not the rest of the integer. Other checks that follow are the same.

zaerl commented on src/parsers/xml-processor/class-importentity.php:
```diff
@@ -0,0 +1,66 @@
+	const TYPE_USER         = 'user';
+	const TYPE_SITE_OPTION  = 'site_option';
+
+	const POST_FIELDS = array(
+		'post_title',
+		'link',
+		'guid',
+		'post_excerpt',
+		'post_published_at',
+		'post_author',
+		'post_content',
+		'post_excerpt',
+		'post_id',
+		'post_status',
+		'post_date',
+		'post_date_gmt',
+		'post_modified',
+		'post_modified_gmt',
+		'comment_status',
+		'ping_status',
+		'post_name',
+		'post_parent',
+		'menu_order',
+		'post_type',
+		'post_password',
+		'is_sticky',
+		'attachment_url',
+	);
```
I'm not very sure about this list here. Where did we get this? Shouldn't it be the list found on `WXREntityReader::__construct`?

adamziel replied:
I don't think this constant is even used in this PR, I'll remove it

zaerl commented on src/parsers/xml-processor/class-wxrentityreader.php:
```diff
@@ -0,0 +1,984 @@
+		// Unfortunately, different WXR files defined the `wp` namespace in a different way.
+		// Folks use a mixture of HTTP vs HTTPS protocols and version numbers. We must
+		// account for all possible options to parse these documents correctly.
+		$wxr_namespaces       = array(
+			'http://wordpress.org/export/1.0/',
+			'https://wordpress.org/export/1.0/',
+			'http://wordpress.org/export/1.1/',
+			'https://wordpress.org/export/1.1/',
+			'http://wordpress.org/export/1.2/',
+			'https://wordpress.org/export/1.2/',
```
I like the approach here, even if it generates `x6` more data. Great work.

adamziel commented on src/parsers/xml-processor/utf8.php:
```diff
@@ -0,0 +1,363 @@
+ * @since {WP_VERSION}
+ *
+ */
+function utf8_code_point_count( $bytes, &$first_error_byte_at = null ) {
```
Renamed, thank you

zaerl commented on src/parsers/xml-processor/class-xmlunsupportedexception.php:
```diff
@@ -0,0 +1,97 @@
+ *
+ * This class is designed for internal use by the XML processor.
+ *
+ * The XML API aims to operate in compliance with the XML5
```
Did you mean XML 1.1?

adamziel replied:
Ah, I meant XML 1.0. Not sure where 5 came from

zaerl commented:
I really like this, some fresh air is much needed in our (legacy?) WordPress importer plugin. We can land this.

adamziel commented on src/parsers/xml-processor/utf8.php:
```diff
@@ -0,0 +1,363 @@
+	 *
+	 * @see http://bjoern.hoehrmann.de/utf-8/decoder/dfa/
+	 */
+	static $state_table = (
+		"\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00" .
+		"\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00" .
+		"\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00" .
+		"\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00" .
+		"\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x09\x09\x09\x09\x09\x09\x09\x09\x09\x09\x09\x09\x09\x09\x09\x09" .
+		"\x07\x07\x07\x07\x07\x07\x07\x07\x07\x07\x07\x07\x07\x07\x07\x07\x07\x07\x07\x07\x07\x07\x07\x07\x07\x07\x07\x07\x07\x07\x07\x07" .
+		"\x08\x08\x02\x02\x02\x02\x02\x02\x02\x02\x02\x02\x02\x02\x02\x02\x02\x02\x02\x02\x02\x02\x02\x02\x02\x02\x02\x02\x02\x02\x02\x02" .
+		"\x10\x03\x03\x03\x03\x03\x03\x03\x03\x03\x03\x03\x03\x04\x03\x03" .
+		"\x11\x06\x06\x06\x05\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08" .
+		"\x00\x01\x02\x03\x05\x08\x07\x01\x01\x01\x04\x06\x01\x01\x01\x01" .
+		"\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x00\x01\x01\x01\x01\x01\x00\x01\x00\x01\x01\x01\x01\x01\x01" .
+		"\x01\x02\x01\x01\x01\x01\x01\x02\x01\x02\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x02\x01\x01\x01\x01\x01\x01\x01\x01" .
+		"\x01\x02\x01\x01\x01\x01\x01\x01\x01\x02\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x03\x01\x03\x01\x01\x01\x01\x01\x01" .
+		"\x01\x03\x01\x01\x01\x01\x01\x03\x01\x03\x01\x01\x01\x01\x01\x01\x01\x03\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01"
```
I think @dmsnell is tweaking that entire code in https://core.trac.wordpress.org/ticket/63863 – I'll leave this table as it is for now and then backport any improvements Dennis makes in his patch.

adamziel commented on src/parsers/xml-processor/utf8.php:
```diff
@@ -0,0 +1,363 @@
+}
+
+/**
+ * Extract a slice of a text by code point, where invalid byte seuqences count
```
Reworded, thanks
